src/reflection-type.ts

```diff
@@ -173,19 +176,46 @@ export class Traversal<Root, Leaf> {
     assert(this.root.isA && this.root.isA(value), 'Root value does not match expected root type');
   }
 
-  private createPathsToName(from: NamedTypeDefinition<unknown>, leaf: NamedTypeDefinition<Leaf>) {
-    calculateReverseReach(new Set(), this.cache, from, leaf, false);
-    const found = this.cache.get(leaf);
+  private addAncestor(
+    ancestors: Map<NamedTypeDefinition<unknown>, Path[]>,
+    ancestor: NamedTypeDefinition<unknown>,
+    path: Path
+  ) {
+    const pathsToAncestor = ancestors.get(ancestor);
+    if (!pathsToAncestor) {
+      return ancestors.set(ancestor, [path]);
+    }
+    pathsToAncestor.push(path);
+    return ancestors;
+  }
+
+  private ancestorNamedObjects(
+    target: NamedTypeDefinition<unknown>
+  ): Map<NamedTypeDefinition<unknown>, Path[]> {
+    const found = this.cache.get(target);
     if (!found) {
       return assert.fail('no path to target');
     }
-    for (const parents of found.values()) {
-      // we can only find objects with isA during traversal
-      // todo: we could relax this a bit by finding the unambiguous parent of parent if there is one
-      parents.forEach(parent =>
-        assert(parent.isA, 'nearest containing named thing is not an object ' + parent.name)
-      );
+    const allAncestors: Map<NamedTypeDefinition<unknown>, Path[]> = new Map();
+    for (const [pathStr, parents] of found.entries()) {
+      const path: Path = JSON.parse(pathStr);
+      parents.forEach(ancestor => {
+        if (ancestor.isA) {
+          this.addAncestor(allAncestors, ancestor, path);
+        } else if (['array', 'named'].indexOf(ancestor.definition.type) >= 0) {
+          for (const [namedObjectAncestor, paths] of this.ancestorNamedObjects(
+            ancestor
+          ).entries()) {
+            paths.forEach(pathFromAncestor =>
+              this.addAncestor(allAncestors, namedObjectAncestor, [...pathFromAncestor, ...path])
+            );
+          }
+        } else {
+          assert.fail('nearest containing named thing is not an object: ' + ancestor.name);
+        }
+      });
     }
+    return this.dedupePaths(allAncestors);
```

Comment:
What's the point of deduping the paths? In which situation would we get duplicate ones?